client/namespace.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package client
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/keys"
+	"github.com/cockroachdb/cockroach/proto"
+)
+
+// TODO(pmattis):
+//
+// - This file contains the experimental Cockroach table-based interface. The
+//   API will eventually be moved into db.go, but is collected here during
+//   initial development. Much of the implementation will eventually wind up on
+//   the server using new table-based requests to perform operations.
+//
+// - Add support for namespaces. Currently namespace ID 0 is hardcoded.
+
+// CreateTable creates a table from the specified schema. Table creation will
+// fail if the table name is already in use.
+func (db *DB) CreateTable(schema proto.TableSchema) error {
+	desc := proto.TableDescFromSchema(schema)
+	if err := proto.ValidateTableDesc(desc); err != nil {
+		return err
+	}
+
+	nameKey := keys.MakeNameMetadataKey(0, desc.Name)
+
+	// This isn't strictly necessary as the conditional put below will fail if
+	// the key already exists, but it seems good to avoid the table ID allocation
+	// in most cases when the table already exists.
+	if gr, err := db.Get(nameKey); err != nil {
+		return err
+	} else if gr.Exists() {
+		return fmt.Errorf("table \"%s\" already exists", desc.Name)
+	}
+
+	ir, err := db.Inc(keys.DescIDGenerator, 1)
+	if err != nil {
+		return err
+	}
+	desc.ID = uint32(ir.ValueInt() - 1)
```

Comment:
Perhaps, but our other `*IDGenerator` keys store the next ID to allocate, not the last ID allocated. I'd prefer consistency with the other ID generator keys.